### ISSUE TITLE
BART extract_features: memory reduction options + validity of running batch mode

### ISSUE THREAD

**Issue by seraphinatarrant** [pytorch/fairseq#1919]
I'm trying to use BART as an encoder of another system using `bart.extract_features`. But I'm running into a couple issues that I'd like advice trying to find workarounds for.

1) BART (bart-large) takes up a lot of memory which makes me end up getting OOM (even in eval mode) when I try to train a separate (very lightweight) model at the same time. 
I am using `bart.eval() bart.half() bart.cuda()` and limiting the maximum sequence length to encode (to 200), but it's still an issue. The only function I am calling is bart.extract_features.
So questions:
* Is there anything else I can do memory saving wise?
* I thought about trying to rip out parts of the model that I shouldn't need - like the entire decoder. But when I experimented with this it seems like the extract_features code actually accesses a number of different decoder functions/properties. Any suggestions in this regard?

I'm just using one GPU with 11GB (RTX 2080) but my batch size is tiny (literally 2 at the moment) so I'm not sure parallelising would do much.

2) I have tried running `bart.extract_features` on each sample individually, and also on a batch of samples. I've noticed that if I then compare the stacked output of individually running a set of samples and of running a batch, `torch.eq(a,b)` is false but if I print out the output the tensors are the same.
This is because at some precision (~8 digits, based on a couple examples) the tensor element values _are_ different (and aren't appearing in print by default). 

Can you shed any light on why this is, and whether I need to worry about it? It seems quite small but I also don't know why it is and haven't checked how much variance there is in the differences.

Code snippets for single sample vs. batch are:
```
batch = collate_tokens([bart.encode(s) for s in text], pad_idx=1)
b_t = bart.extract_features(batch)
s_t = torch.stack([bart.extract_features(x).squeeze(0) for x in batch])
```

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.3.1
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7.6
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: bart-large
 - Any other relevant information:

**Comment by tuhinjubcse** [comment]
@myleott Can you please help us here . We are stuck on this

**Comment by seraphinatarrant** [comment]
ok I worked out basically what the major issue is - that `extract_features` was accumulating gradients, whereas if I run other inference code the generation is wrapping in `no_grad` so it's not a problem.

**Comment by huihuifan** [comment]
@ngoyal2707
